teps/0131-larger-results-via-workspace.md

```diff
@@ -0,0 +1,303 @@
+The goal of [TEP-0086][tep-0086] is to support larger `Results` beyond the current size limits. TEP-0086 has many
+alternatives, this TEP proposes implementing one of the alternatives i.e. larger results via workspace. This
+allows us to support larger `Results` to be stored outside of etcd, thus alleviating the practical limit imposed by etcd. 
+
+## Motivation
+
+> As a general rule-of-thumb, if a `Result` needs to be larger than a kilobyte, you should likely use a `Workspace`
+
+1. Situations where size of the result is hard to predict and one would like tekton (or an external system) to determine the most optimal way of storing. 
+  
+  One such situation happens in our use case of Kubeflow pipelines with tekton as backend(kfp-tekton), where an end user writes python code and then the kubeflow pipelines compiler turns it into a tekton pipeline. The end user has no idea, what tekton is capable of and what are its limitations in terms of result size storage. 
+
+2. Results on workspace gives us the flexibility of storage on an external system. Workspace are *not* necessarily backed by filesystem that are optimal for storing large objects, they may be backed by a Software as a storage. We are encouraged by this approach as we currently use Software as storage to store artifacts produced as results from tasks in tekton. 
+
+3. Alleviate the dependence on etcd and its storage limits and without compormising largely on performance benefits. 
+  Etcd imposes a limit of 1.5 mb per object, and is very performant. One could also think of it as a distributed database. A workspace on the other hand, can plugin a database of desired performance characteristics - if there is CSI driver available for it. In other words, achieve desired storage characteristics for storing results of any size. 
+
+4. Workspaces are seen as files storage. A Workspace with underlying PVC backed by a CSI driver for cloud object store is one such example, but it can be some very high performance feature rich proprietory software as a storage solution too.
+
+5. Etcd limit issue can be further aggravated when number of pipelines resources are large and since etcd storage is shared across all kubernetes resources it is a precious resource in most production systems.
+
+
+Q. Why not just use workspace for everything then?
+
+Take advantages of following features of `Results`, 
+1. Simplifying generated yamls. 
+2. ability to evaluate in when expressions. 
+
+We could have a config parameter `result-size-threshold`, which decide if the size is under that threshold, then we allow it to be evaluated as When experessions.
+
+Results with workspace can also be configured tekton wide.
+
+### Goals
+
+ 1. Store outside of etcd as it is a precious shared resource.
+ 2. Allow results with any size, i.e. let an external system decide how results can be stored most optimally.
+
+### Non-Goals
+
+1. Blurr the boundary between workspace and results.
```

Comment:
This is where I think it is kind-of hard. Results, in tekton, so far, are simple, limited in size strings (just like some metadata like names, …). This proposes to change this and "blurry" that result notion and an kind of "artifact" notion (json test report, seriff files, …).

Reply:
I am open to calling it artifact!

Reply:
@ScrapCodes we're defining tekton artifacts in this doc - please take a look 

https://docs.google.com/document/d/13qxkpC0m5zTEWJ_oprA2ZO9D0wj_Q4xyz_UMiJchujU/edit?usp=sharing&resourcekey=0-13bMrFgBumiIkEEGBQRfVw

Reply:
> 1. Blurr the boundary between workspace and results.

This is marked as a *non-goal* i.e. there is no intention of this blur happening.
Whether we call this functionality artifact or "large result", as long as we provide an API for users to use this, I think it won't be mixed up with regular results.

In my view:
- Results are small-ish bit of data, passed along by value.
- Artifacts are larger chunks of data, passed along by reference. Hopefully, we'll add metadata to these and getter interesting features in terms of attestations, eventing and more.